modules/dists/BlockDist.chpl

```diff
@@ -1767,35 +1772,42 @@ proc BlockArr.doiScan(op, dom) where (rank == 1) &&
       if debugBlockScan then
         writeln(locid, ": ", (numTasks, rngs, state, tot));
 
+      // Create a local ready var and store it back on the initiating
+      // locale so it can notify us
+      var myOutputReady = new unmanaged BoxedSync(bool)?;
+      outputReady$[locid] = myOutputReady;
+
       // save our local scan total away and signal that it's ready
-      elemPerLoc[1] = tot;
-      inputReady$[1].writeEF(true);
+      elemPerLoc[locid] = tot;
+      inputReady$[locid].writeEF(true);
 
       // the "first" locale scans the per-locale contributions as they
       // become ready
-      if (locid == dom.dist.targetLocDom.lowBound) {
+      if (locid == dom.dist.targetLocDom.lowBound) then on elemPerLoc {
         const metaop = op.clone();
 
         var next: resType = metaop.identity;
         for locid in dom.dist.targetLocDom {
-          const targetloc = targetLocs[locid];
-          const locready = inputReady$.replicand(targetloc)[1].readFE();
+          const locready = inputReady$[locid].readFE();
 
-          // store the scan value and mark that it's ready
-          ref locVal = elemPerLoc.replicand(targetloc)[1];
+          // store the scan value
+          ref locVal = elemPerLoc[locid];
           locVal <=> next;
-          outputReady$.replicand(targetloc)[1].writeEF(true);
 
           // accumulate to prep for the next iteration
           metaop.accumulateOntoState(next, locVal);
         }
+        // Mark that scan values are ready
+        coforall ready in outputReady$ do on ready {
+          ready!.writeEF(true);
+        }
```

Comment:
Using this `coforall` rather than a per-locale signal as we used to do on 1787 using `begin on` (say) made me wonder what the tradeoffs would be.  Guessing:
* `begin` + `on` might spin up a local task rather than simply creating a remote one?  (and worse, on the current core?)
* we'd need a `sync` to know it was safe to proceed, which may not be as optimal as far as pre-populating it with the end-count goes?

Reply:
Just wanted to check and I do see slightly worse perf with begin+on:

```diff
diff --git a/modules/dists/BlockDist.chpl b/modules/dists/BlockDist.chpl
index 5c9627bb1a..31456c5e8e 100644
--- a/modules/dists/BlockDist.chpl
+++ b/modules/dists/BlockDist.chpl
@@ -1787,19 +1787,19 @@ proc BlockArr.doiScan(op, dom) where (rank == 1) &&
         const metaop = op.clone();

         var next: resType = metaop.identity;
+        sync {
         for locid in dom.dist.targetLocDom {
           const locready = inputReady$[locid].readFE();

           // store the scan value
           ref locVal = elemPerLoc[locid];
           locVal <=> next;
+          var ready = outputReady$[locid];
+          begin on ready do ready!.writeEF(true);

           // accumulate to prep for the next iteration
           metaop.accumulateOntoState(next, locVal);
         }
-        // Mark that scan values are ready
-        coforall ready in outputReady$ do on ready {
-          ready!.writeEF(true);
         }
         delete metaop;
       }
```

```
chpl test/scan/scanPerf.chpl --fast --no-cache-remote
./scanPerf --printTiming --printArray=false --n=1024 -nl 512
Scan of 1024 elements took 0.003953 seconds

git apply begin-on-scan.diff
chpl test/scan/scanPerf.chpl --fast --no-cache-remote
./scanPerf --printTiming --printArray=false --n=1024 -nl 512
Scan of 1024 elements took 0.005297 seconds
```